### DISCUSSION TITLE
Ideation and roadmap for Jamulus 4

### DISCUSSION THREAD

**Discussion by ann0see** [jamulussoftware/jamulus#3341] [Ideas]
With the project for Jamulus 4 already existing (https://github.com/jamulussoftware/jamulus/milestone/9), I think it makes sense to define a broader goal roadmap for Jamulus 4.

By definition Jamulus 4 would introduce breaking changes like more IPv6 support and potentially larger scale improvements.

Moreover, this should be a larger release and must be well planned. 

So far I see the following feature:

Architectural changes:
* Fixing fragmentation issues with improved IPv6 support
* Restructuring the GUI to be more decoupled

User visible changes:
* More mobile support
* Improve JSON-RPC support
* Potentially modernized UI glyphs + logo
* Dropping support for legacy builds

Project changes:
* Rethink release process to be quicker

I would like to start a discussion on how we want to make 4.0.0 look like. Please your thoughts about long term goals!

**Comment by dcorson-ticino-com** [comment]
Those are all behind the scenes things that will make things better for some, but hardly a reason for a major version change.
Probably a lot of programming but most users will see no difference, other than maybe their usual UI being changed a bit. 

IMHO the most pressing improvement would be to add uncompressed PCM to have better sound quality.  
That is the first thing I hear from all new users, all the competition has it and it makes a big difference in the sound quality.

(our biggest advantage compared to the competition is that Jamulus never crashes.  Never a break to reboot or people getting lost, which is common for our competition.  Jamulus just works.)

I had an uncompressed version, with a 5th audio quality "best", almost working, but never found what was causing buzzing at times and gave up.  Unfortunately I don't have the time to pick this up again in the near future.

**Reply by ann0see** [reply]
Agree. Sound quality is also a thing I heard about in the past. 
The version change is mostly due to potentially breaking backwards compatibility. 
I agree that the user should also feel that the version change is worth it.

**Comment by pljones** [comment]
> More mobile support

This is dependent upon two separate points:
* Qt support for mobile -- in my experience trying to get touch events working on Android, it wasn't that easy.  We may need to include "Drop Qt5 entirely" in the list of significant architectural changes.
* Mobile-competent developers -- we're a small team and I'm not sure we've got the necessary skills to investigate and resolve issues on a wide range of platforms.  For example, I had _severe_ issues getting the Android build running under Windows and Linux, to the extent I gave up and relied on Github.  That's not a good way to work.

On the Qt6 point above, we may also want to move entirely to Qt6 for a more consistent UI look-and-feel across platforms.  However, getting good UX on mobile and desktop devices will need more than just Qt6, it needs a complete reworking of how the UI components are handled (on different runtime platforms).

**Reply by ann0see** [reply]
> I'm not sure we've got the necessary skills to investigate and resolve issues on a wide range of platforms. For example, I had severe issues getting the Android build running under Windows and Linux, to the extent I gave up and relied on Github.

No. Currently we don't. However, for me I'd assume iOS at least is learnable. 
If we get further with mobile, I'd assume we will get more contributors for those platforms too. 

In any case, the GUI issue shows up in multiple places and needs to be done eventually.

**Reply by pljones** [reply]
I think we also need to review the build processes -- it needs to be easier for a developer for a specific platform to understand how they can proceed locally with development for that platform.  Currently there's probably too much reliance on Github tooling, rather than on platform tooling.  (I don't mind pulling in a Github build platform image, so long as we keep that to current, supported releases of the platform and we can say how to install - by hand, from a version supported by its supplier - each of the needed tools to build the artefacts we build on that image.)

**Reply by ann0see** [reply]
So you're saying we should ensure that each OS has a build script and working + tested setup guides. I think although we aren't there yet fully, it's doable.

**Reply by pljones** [reply]
I'm saying any developer for a platform should be able to develop for that platform without reliance on Github to do the builds.  Not only does this give the developer faster feedback (i.e. they don't need to wait for the Windows ASIO build) by allowing them to test and fix locally, it also enables the Jamulus project to have less dependence upon Github, should there ever come a time when it can no longer be used.

**Reply by ann0see** [reply]
Linux -> Works
macOS -> Works with quirks and script
iOS -> Works
Windows -> Hassle to set up but works with QtCreator as far as I know
Android -> Hassle to set up. Only works with docker container.

**Reply by pljones** [reply]
Yeah, two out of five is what I mean - it's not that good.  And for Android, it needs to work on Windows and Linux based development systems (and MacOS, if it supports Android development), ideally directly inside QtCreator.

(QtCreator on Windows sucks.  I don't know how they've managed to make it run so slowly - it runs as slowly on my shiny new 32core CPU as it did on my old 8core..  They must have single threaded absolutely everything (including syntax checking, highlighting, etc).  I generally use the Linux version remotely desktop'd because of that!)

**Comment by stevie2sheds** [comment]
Multi-channel support would be another great addition for a major version. Currently Jamulus supports a stereo pair meaning that users have to mix in a vocal using patchbay/Blackhole/VoiceMeter or some analog mixer solution. With multichannel interfaces that shouldn't be necessary. Along with uncompressed PCM, being able to send, and therefore record, multiple channels would make Jamulus a very good recording utility.

**Reply by dcorson-ticino-com** [reply]
Multichannel support is a problem for Jamulus.
We could use the two stereo channels separately as two mono channels, but more would mean a total rewrite of the code of both client and server, as I see it.  And probably throw out any possibility of backward compatibility at the same time.
But supporting 2 separate mono channels from a client should be doable.
Just to find someone who finds that feature so important that they are willing to do the development.

**Reply by stevie2sheds** [reply]
Understood. I just felt the topic was searching for those potential, significant changes to warrant a new version. I guess Sonobus can achieve it more easily peer to peer, but that is a great feature it has. As has been highlighted, we use Jamulus because of its relative simplicity and robustness so that shouldn't be lost. I just wish I could mute all the microphones bleeding in (or rather I wish my bandmates would mute their microphones bleeding in)!

**Reply by dcorson-ticino-com** [reply]
If the problem is just muting micros a two channel version would work and would let you mute them.
I think that is doable.  Thanks!

**Reply by pljones** [reply]
A "multi-channel client", in Jamulus terms, would be exactly the same as running multiple instances of the client -- just with the advantage _on Windows ASIO_ of having a single ASIO connection, which some ASIO drivers might find helpful.

It would make multi-platform support even harder, of course - it's hard enough currently keeping the simple view of different platforms' ways of providing audio channels coherent.  To add cross-platform multi-channel support is certainly do-able (in the sense above - instantiate multiple client connections to the server for each audio in mono or stereo pair) but would probably take a few years to achieve.

Given that on all platforms there are alternatives (either via JACK on Windows or the platform itself) using multiple instances of Jamulus already, the time spent on it might not be considered that worthwhile if there were other goals that could be achieved.

**Reply by ann0see** [reply]
Related to: https://github.com/jamulussoftware/jamulus/issues/516

I don't think that we will get multi channel support in Jamulus 4. PRs are still welcome, but this could require a lot of work.

**Comment by rdica** [comment]
Access control functionality.

As I understand it, @Rob-NY has already submitted RPC code updates awaiting approval to be merged that provides new RPC calls to manage access to server by IP address.

**Reply by pljones** [reply]
I'd veto this for any Server using the public Directories.  By definition, such Servers are open to the public.

**Reply by rdica** [reply]
Perhaps I don't understand your definition of open to the public; to be clear, servers that are registered on public directories and firewalling are fine, but having the ability to perform the same function from the jamulus software itself is not, do I understand your position correctly?

**Reply by ann0see** [reply]
Publically registered servers should be open. 
My personal point is that I'd like to have some kind of access control - especially on private servers, but I know that @pljones objects it.

**Reply by pljones** [reply]
> servers that are registered on public directories and firewalling are fine

No, they're not.  We just can't stop it happening.  They're using up a freely provided service that is intended to allow all Jamulus users to find publicly available Servers and making those Servers not available to all.  That's not the intent by which the public Directories are provided.

**Reply by mcfnord** [reply]
I have only seen this feature used in petty, hostile ways that diminish the utility of the mesh network.

**Reply by rdica** [reply]
The potential to have permitted clients utilize the RPC interface to manage access to the server while in a session would be a useful function, and relieve the permitted client of the need to ssh/rdp/www into a server and add a firewall rule, interrupting/delaying playing. (Yes, it would mean running the RPC listener on a WAN interface, not localhost, but access to that could be defined once in a firewall config, sharing the JSON-RPC "secret" to predetermined client(s), inputting that value into a field in the Settings dialog for example)

Adding functionality to the server GUI log dialog box to select then block using the backend RPC interface could prove very useful as well for those that run a server from their home connection and may not be familiar with or permitted to alter their routers/firewalls.

@mcfnord It's not surprising you'd **only** see it that way given you have a personal stake in this as you, and your bots, are blocked on a number of servers. That is a situation of your own making, and doesn't bear on the ACL discussion at large. What you perceive as petty and hostile sounds like a personal issue best left off this discussion.

I see firewalling being used in healthy, productive ways, benefitting player communities and server operators. The utility of the mesh network isn't diminished at all. Directory servers aren't blocked, indexers like explorer.jamulus.io aren't blocked. A user being blocked on a server has no impact whatsoever on the mesh network itself, or the rest of the users using jamulus.

@pljones I can understand your feelings on this.
From my point of view I see it as Google telling Apache/Nginx they cannot have any kind of ACL control else they won't index you.
I don't agree with your position, but I respect it.

Thanks for the explanation, it's much appreciated, and thanks for your consideration.

**Reply by ann0see** [reply]
> It's not surprising you'd only see it that way given you have a personal stake in this as you, and your bots, are blocked on a number of servers. That is a situation of your own making, and doesn't bear on the ACL discussion at large. What you perceive as petty and hostile sounds like a personal issue best left off this discussion.

Off topic.

**Reply by pljones** [reply]
> From my point of view I see it as Google telling Apache/Nginx they cannot have any kind of ACL control else they won't index you.
I don't agree with your position, but I respect it.

I don't think you can compare a corporation like Google with the providers of the public Directories.  That's stretching things a long way.

**Reply by ann0see** [reply]
Related to: https://github.com/orgs/jamulussoftware/discussions/1057 Feel free to discuss further in this discussion. I think even though it's possible to implement this, we'd need to agree on what is acceptable and what isn't.

**Comment by chrisrimple** [comment]
Random ideas to consider, some of which may have been suggested long ago but deemed too hard or inappropriate at the time...

- Make it easier for less-technical users to implement a headless server (presumably on rented server space).  There are plenty of instructions out there, and even some pre-packaging of deployments, but the user still needs to edit the configuration manually (often using a Linux editor with which they're probably unfamiliar).  I'm thinking of no-code solutions like a "built a server" UI that allows selection of options and spits out a package that can then be deployed with a single Linux command, or even a "copy a server" solution.
- Make it easier to run multiple instances of the client, on Mac at least.  Unlike Windows, where the user can specify a unique INI file for each instance in a Shortcut, the Mac user has no easy solution.  If Jamulus looked for an INI file based on the application name, rather than a default name, a Mac user could copy the client and rename it (Jamulus1.app, Jamulus2.app, etc.), resulting in distinct INI files for each instance when they are run.
- In Settings > Audio, break the Audio Device dropdown into two, one for IN and one for OUT, and place the OUT above the Output Channel Mapping.  No change to functionality, just a usability improvement.
- Make Directory Servers more robust or unnecessary, to prevent any possibility of downtime.  Maybe that means having backup Directory Servers that are always running with automatic failover, or having a list of backup Directory Servers that a Client can use when a primary Directory Server doesn't respond within a set time.
- Improve the UX for the user information that appears below each mixer fader.  For example, Alias/Name is limited to 16 characters, but the word wrap isn't smart and always breaks after character 8, even if that's mid-word.  The Instrument and Country/Region are positioned to the left, reducing the space available for text (Alias/Name), when they could be positioned above or below.

**Reply by pljones** [reply]
The server UI could do with some updating for the Jamulus Server-specific things, I'd strongly agree.  However, platform-specific installers that handle deployment, beyond what we've already got are unlikely, I'd say -- each one requires separate development and maintenance.

Multiple installs for the client - even on Windows - could be improved.  It's not _nice_ having to edit a shortcut.  There are things like the Windows installer not differentiating between the ASIO and Jack installs, so installing both means only the last install run is remembered.  So yes, getting the installers better would be good.  But, as with the first point, it's not within Jamulus, it's something that would need to be handled on a platform specific basis separately from Jamulus development.

Could you post a screenshot for the third point, pointing out where the problem is in context?

All the public Directories are on a "best endeavours" basis by volunteers.  If we had more volunteers willing to run them, it could be improved.  But having fail-over would mean some kind of monitoring service running somewhere (potentially multiple instances - three and if two shout "down!" you've a problem) to decide to switch the DNS entries (on which we rely).  Noting that the DNS entries have fairly short expiry, IIRC (hours rather than months).  A soft-closed Directory saves state, which it could ship off site, so that if it didn't recover, the safe copy could be used for faster recovery (we've done this in the past, manually).  At the moment, that copy would be going stale immediately but a refresh could be done (bounce the Directory once an hour - you'd be unlikely to notice the outage - and ship the fresh save file).  Again, this isn't Jamulus we're talking about.

The word wrap on the Alias/Name has actually had quite a lot of work on it.  If anyone wants to spend more time and contribute a patch, they're always welcome.  The layout has to be adaptable to the three different formats, too, plus allow for the stacked view mixer.  There isn't much UI area it's got available - really, taking up more width than the fader is going to hurt quite a lot.

**Reply by pljones** [reply]
And, on MacOS, it doesn't sound too bad:
https://www.makeuseof.com/tag/run-multiple-copies-app-mac/#actually-run-multiple-copies-of-an-app
From what I understand of that, there's no alternative and it's pretty much exactly the same as multiple shortcuts on Windows.

**Reply by pljones** [reply]
@gilgongo @ann0see if someone with a Mac can confirm the above, we could add a section to https://jamulus.io/wiki/Tips-Tricks-More about running multiple instances on Windows, Mac and Linux.

**Reply by chrisrimple** [reply]
The details on using a Script to run multiple instances on Mac are accurate.  BUT using a Script is only necessary if you care about having separate INI files for each instance.  A simpler solution on Mac for the person who only needs 2-3 instances (and one that I detail in [my doc](https://docs.google.com/document/d/1smcvsxdaaViPQvGMQHmah_6BQeqowhmGSFMHfnlY2FI/edit?usp=share_link), which is linked from https://jamulus.io/wiki/Tips-Tricks-More already) is to make a copy of the Jamulus application (Jamulus.app in the Applications folder) with a different name, then run both (or many) - as each is launched, it will have the (single) INI settings, but then can be manually updated before connecting to a server.  When quitting the app, the INI file will be overwritten by the settings from the last instance running.

**Reply by pljones** [reply]
Is it simpler to lose settings you thought you'd changed?  As far as I can see, without passing the ini filename explicitly, there doesn't appear to be a way for a Mac app to _know_ where it's run from and guess either it's name or directory.  I could (likely) be wrong - I'm basing this on uninformed Google searches that might not even be asking the right question.

**Reply by chrisrimple** [reply]
You're correct: MacOS doesn't care about the location of the app - it looks for the INI file in a pre-defined location regardless of the app location.  So scripts are the only way in which a user can specify a unique INI file (whether running multiple instances or not).

Yes, you could lose settings you thought you'd changed, but the harm isn't significant.  Worst case, you run Jamulus and have the wrong Audio Device selected, the wrong Name/Icon/etc., and maybe a couple other settings that you have to update before Connecting.  It only takes a moment to fix all that.  When the WorldJam was active (during Covid), I was regularly running 3 instances of Jamulus without using scripts, and just ensured that I quit last the app instance that had settings I wanted to retain in the INI.

An advantage to NOT using multiple INIs (via scripts) is that settings related to Solo, Mute, etc. are true for all app instances, rather than being maintained separately.  This is particularly handy when all the app instances will be connected to the same server.

Keep in mind that only more-technically-capable users are likely to be running multiple instances, so they're less likely to be bothered by the need to update some settings when running the app.

A v4 feature might be to add an INI Location field to Jamulus, allowing the user to specify the file to be loaded, with a Load button.  Then a user could run Jamulus1.app and Load their INI1 file, followed by running Jamulus2.app and Load their INI2 file (which they made by copying INI1).  Over time, the two INI files will diverge, but the user with multiple instances won't need to use scripts to access them.

**Reply by pljones** [reply]
Without a substantial rewrite of the application, reading the inifile at anything other than application start up won't happen.  There have been at least two attempts at rewriting the inifile handling, both of which failed to produce an improvement, so far, so I wouldn't hold your breath on "load whilst running" happening soon.

(Jamulus determines if it's "client" or "server", instantiates the settings object for client or server from the appropriate inifile and then instantiates a client or server, passing it the settings object.  It would need to re-instantiate the client or server after looping back out to re-read the updated inifile if the location were changed "on the fly" - and, if the location had been passed on the command line, it would still use that location instead... unless something even more confusing happened...)

**Reply by ann0see** [reply]
@chrisrimple Could you please open an issue on the mixer board change and sum up the discussion here? I think this could be part of Jamulus 4 - if we are clear on what to improve.

**Comment by chrisrimple** [comment]
![Screen Shot 2024-08-15 at 9 32 58 AM](https://github.com/user-attachments/assets/7abaac39-0846-4822-9909-e2a78299a914)

See the Audio Device dropdown, which has both IN and OUT in a single list.  I propose that they be broken up, with an IN dropdown above the Input Channel Mapping controls and an OUT dropdown above the Output Channel Mapping controls.

**Reply by ann0see** [reply]
Yes. There's an issue about that but it's not trivial to implement. But yes, this could be part of Jamulus 4

**Reply by pljones** [reply]
As I understand it, currently:
* Audio Device specifies which "hardware" devices are declared by the operating system.  The content isn't under Jamulus's control.
* The Channel Mappings specify which channels are declared by the selected hardware device.  The content isn't under Jamulus's control.

We could add a choice for each of the Input and Output mappings (to make it even more complicated than it is now):
- [ ] Consecutive odd+even channels form a stereo pair (single mapping labelled "Stereo pair")
- [ ] Consecutive channels (free) form a stereo pair (single mapping labelled "Stereo pair")
- [ ] Free selection of stereo pair (i.e. what's offered currently when a hardware device supports more than two channels)
- [ ] Mono channel  (single mapping labelled "Mono")

I'd also like to see removal of the "vanishing mapping" when there's only two channels (like with Jack).

**Reply by ann0see** [reply]
> Audio Device specifies which "hardware" devices are declared by the operating system. The content isn't under Jamulus's control.

This does not fully hold true for macOS and iOS. In macOS the standard audio system is capable enough. The user selects the correct input output tuple from the Cartesian product of input and output devices. And this is as user unfriendly as it sounds.

**Reply by pljones** [reply]
OK, so we should say "Core Audio" for the audio device and lock it?  Presumably there's a reason it's like it is...  Can you take inputs across different hardware inputs, for example, and retain buffer synchronisation?

**Reply by ann0see** [reply]
> Can you take inputs across different hardware inputs, for example, and retain buffer synchronisation?

Not sure. But there's a clock synchronization toggle in audio/midi setup on macOS - the application where you configure advanced audio settings.

**Reply by pljones** [reply]
Hopefully the core audio api would support an application turning that on, if needed, too.  Sounds like what we'd need, I think.

**Reply by ann0see** [reply]
Related to: https://github.com/jamulussoftware/jamulus/issues/3216

@pljones can you copy your suggestion into this issue please?

**Comment by pcar75** [comment]
Some Jamulus' sessions are enviably professional, hosting amazing musicians. To me, it would make sense to have the focused server able to record better quality audio (no lossy audio stream as alluded before but also adjustable sample rates 48 to 96 mHz and bit depths of 24 or 32).
This might entail that 
1) the client and server has the capacity to generate and process better audio (impact to code; Opus library? );
2) there is a facility to gauge if a server's processing capacity is sufficient to use better quality streams (i.e. some kind of Jamulus benchmark tool ?).

To developers and supporters , my sincerest thanks.

**Reply by pljones** [reply]
The server already has the ability to record multi-stream audio.

The client doesn't _receive_ lossless audio or multi-stream audio (and, to remain "Jamulus" it never will get mulit-stream).

**Reply by pcar75** [reply]
@pljones 
Bonjour,  thanks but I already understood this info ... Since this is "**ideation**" for the future, I will try to be more explicit.

The idea is to enable a mix engineer/enthousiast to have the best canvas possible.  In order to avoid unnecessary noise (e.g. single or multiple aliasing, dithering, down/up conversion, etc. in the server process of recording) my understanding is to start from the best tracks quality possible, which means getting and recording lossless, 24 or higher bit depth, possibly higher than 48 mHz sample rate.

On the recording side, starting at the source, a client would send it's sound lossless at say 24bit / 96 mHz.  The server could record this a  .wav with 24bit / 96 mHz or at least the best quality among the tracks received.
On the live side, as usual, the rough stereo (mono?) mix would be sent to each clients, possibly along it's capability (lossless and at local bit depth and sampling rate).

At least, beyond lossless stream, the bit depth could be enhanced to 24bits.

Note: 
_"(...) The client doesn't _receive_ lossless audio  (...)"_    unless I'm mistaken the client does not currently _send_ lossless audio either ...

**Reply by pljones** [reply]
I don't think I'm getting what you're after at all.

First, I'll just say the 16bit 48KHz audio is likely to be part of what Jamulus is for a while.  It keeps the clients and servers in sync without the server having to do too much work resampling (and thus losing more audio signal to noise).

The mixer you see in Jamulus is sending instructions to the server, you understand.  The server does all the mixing for all the connected client channels.  Up to 150 stereo channels in, 150 stereo channels out, currently.

Each of those up to 150 clients only sends or receives a single (mono or) stereo pair.  So the client has nothing to do mixing with.  The audio data is only in the 150 separate streams on the input side at the server before the mix happens.  That's where the server recorder currently captures the data.

Currently, all audio gets compressed, yes.  There's already a patch around, I believe, to work with PCM data in place of OPUS compressed data.  That's not too hard, and can be done in a backwards compatible way (allowing existing clients to connect to a PCM+OPUS server and choose whether to use PCM or OPUS data).

The issue isn't whether it's compressed or not.  It's that the Jamulus server _is_ the mixer.  That's how Jamulus works.  If that's not what the server is doing and you want peer to peer audio, then you don't want to be using Jamulus.  If you want to work with the server recordings, that's fine - Audacity and Reaper projects are both plain text files and the audio is 16bit LPCM WAVs (and, as I say, that's likely not to improve in V4 unless someone wants to show a V3-compatible audio engine for the server that can do it).

**Reply by pcar75** [reply]
Lossless will be nice for those with accomodating bandwith , but what beyond ?

In case of different bit-depth/sampling (characteristics) of inputs, I see the point that with backward compatibility ("don't break anything" rule) a server could not mix different characteristics streams without resampling some and possibly "...losing more audio signal to noise)...".  But even legacy support is bound to have its limits (Jamulus 1x , 2x , support for Mac pre-X , Windows pre-XP ? ).

I doubt Jamulus is stuck at 16b/48mHz forever ( or at least all it's future existence, which I hope very long :-) so eventually Jamulus will evolve to use better audio quality ; channels/depth/sampling or whatever better ( AD/DA or else ) standards will come along and become norms.

So maybe this is a suggestion for version N roadmap then, not 4.x.
Thanks for the discussion.

**Reply by pljones** [reply]
Working on the basis people generally update the client, we can look at Jamulus Explorer to see what the rule for the server is.

The oldest looks to be is 3.4.2.  I think that may have been the first that required registering with a directory - rather than _needing_ to be set on the directory command line - and the server providing its version number.  However, the oldest tag in git is 3.4.3 (2018-05-17).  3.4.4 (2020-03-25) brought server-side recording

... and then Covid hit!  During Covid, the previous quarterly-ish release cadence picked up to once or twice a month for a while.

3.4.5 (2020-04-04) upped the number of clients to 50; 3.4.6 (2020-04-09) added the mixer channel _meters_ for visual reference; 3.4.7 (2020-04-11) added a second directory, as Covid had flooded the original one.

3.5.0 (2020-04-15) allowed the client to choose which directory provided the server list - but not in the connect dialog; 3.5.3 added multi-lingual support and dropped Qt4 because of that, IIRC; it also added the multi-channel audio settings support, plus recorder improvements (with a fix in 3.5.4) and putting the directory dropdown in the right place.

3.5.11 (2020-09-19) finally brought update checking - so we can assume the may be both clients and servers out there before that but that clients, at least, should be getting notified of any updates.  However: the check is only on start up and a "no GUI" (headless) Jamulus instance only logs the notice to stdout, it can't issue a pop-up, hence the few servers older than that.

3.6.0 (2020-10-25) dropped the ability for a directory to pre-register servers on start up at all in favour of server registration.  Hence _most_ servers are 3.6.0 or newer.  Most of the changes during this period (April 2020 to December 2020) were usability improvements rather than architectural, plus bug fixes.

3.7.0 (2021-03-17) rewrote the build system and new directories on a new domain were introduced - both major changes.  This was also the first release after Volker passed on support of Jamulus.  (This was the only 3.7.x release.)

3.8.0 (2021-06-02) made a breaking change to the server - a truly breaking one, meaning clients from before Feb 2013 (3.3.0) could no longer connect as they didn't not start sending audio before hearing back from the server.  There was a lot of soul-searching.  It also made Qt 5.12.2 the main platform, introducing the "legacy" 5.9.9 build to continue supporting macOS 10.10 to 10.12 (July 2017 to September 2019), again, trying hard to maintain support for legacy platforms.  It also featured some significant UI changes.  3.8.1 (2021-10-23) and 3.8.2 (2022-02-20) followed with minor changes, plus improved experimental iOS and Android support.  3.8.1 slipped in the IPv6 changes, too.

3.9.0 (2022-07-29) introduced the experimental (still) JSON-RPC interface, gaining the x.y.0 bump.  Lots of minor GUI, tooling and packaging changes and Qt6, too!  3.9.1 introduced Qt5 compatiblity back into the ISO country code handling to enable Qt5 and Qt6-based clients to understand each other.  Plus sundry usability improvements.

3.10.0 (2023-09-03) Dropped pre-Windows 10 64bit support with the move to Qt 6.5.2 (use 32bit as "legacy" build).  MacOs legacy now built on 13.1 which isn't officially supported by Qt 5.9.9, so macOS "legacy" is looking close to end of life, too (flagged as deprecated in the release notes).  Otherwise mostly internal changes.

The coming release is pretty much more of the same - internal changes, trying to keep up with dependencies, plus bug fixes and some small feature improvements.

But anything that looks like _breaking support_ for older versions has to be looked at carefully.  We still have that 3.4.2 server.  Clients prior to 3.8.0 could still be using any server older than 3.8.0, too (but are not supported on newer ones).

**Reply by dcorson-ticino-com** [reply]
@pljones could you point me to that PCM patch please. With that I could probably have a working PCM version quickly.
I never got the PCM on the server right, the rest was working as far as I remember (I think I was based on 3.8.X).
We should, of course, discuss how that new PCM quality should present.
What I made was:
a 4th Audio quality "Best" which chose PCM or "High" quality according to the capability of the server.  The actual quality being used was displayed in the settings window. Server capability determined from the server sw version.

**Reply by pljones** [reply]
I think it was probably what you were working on :)

**Comment by ann0see** [comment]
I'd like to close this discussion after the release of 3.11.0 and open a new one for deciding and prioritizing. 
This doesn't mean that the changes here won't happen, just what and who commits for which changes.

**Comment by ann0see** [comment]
To sum up the intermediate state:

I think the top requested Jamulus 4 feature is improved audio quality.

Afterwards thinking about multi channel support, ACL and the other mentioned features.

**Reply by stevie2sheds** [reply]
One more comment is that there is merit to include lesser sample rates (44.1kHz) as well if the intention is for options. I've a friend with this device that would have been perfect for Jamulus use but it only supported 44.1 kHz [https://www.boss.info/uk/products/br-800/](https://www.boss.info/uk/products/br-800/) 

I'd happily use a 44.1 kHz server for online rehearsing.

**Reply by ann0see** [reply]
The correct solution wouldn't be a 44.1 server IMO but resampling. Otherwise it breaks compatibility. 

https://github.com/jamulussoftware/jamulus/issues/15

**Reply by pljones** [reply]
You can use ASIO4ALL to resample on Windows: https://asio4all.org/always-resample-44-1-48-khz/

**Reply by ann0see** [reply]
But ASIO4All is tricky.

**Reply by pljones** [reply]
There are also other solutions outside Jamulus to do resampling.  (Personally, I'd drop the built-in reverb, too...)

**Reply by ann0see** [reply]
The reverb should probably go into advanced settings first. But I'm already seeing negative community reactions if we remove it. 
Best case would be to allow plug ins instead of the inbuilt reverb. Then everyone will be happy.

**Reply by pljones** [reply]
I guess adding CLAP support might not be a bad thing, if feasible.  It should be possible on both the input chain and output chain, I'd think, without any investigation.

**Reply by ann0see** [reply]
From some quick research it seems nice but since it's "new" there might be some quirks.

**Reply by stevie2sheds** [reply]
Thanks folks. The logic in my thought/idea was that everyone can switch down to the lowest setting, if it were available, without resampling, but not every device can easily switch up. For audio quality in general the input from our band members is so variable with artifacts, ground loops, cats, children, network glitches etc that sample rates and bit depths are pretty much irrelevant! Obviously having the option to go better for recording and overall audio improvement is something to be strived for, but low latency and simplicity rightly trumps premium quality audio imo.

+1 on not requiring reverb - as noted that's either a plugin, but ideally should already be there in the source if olk want it. That said it is a feature that's used often by vocals so I think there'd be upset if it were lost.

**Reply by ann0see** [reply]
@pljones I think it's worth opening an issue on CLAP support.
VST is still somewhat related (https://github.com/orgs/jamulussoftware/discussions/2458), even though this is not plugin support in Jamulus directly, but rather Jamulus as VST plugin.

**Comment by ann0see** [comment]
I will close and summarize this discussion tonight due to the release of 3.11.0

Then we can start linking issues and planning further steps. Please add final thoughts if needed. Feel free to find and link or create new issues.

**Comment by ann0see** [comment]
To sum this up:

We have 4 larger feature requests:
1. Provide uncompressed/higher quality audio.
2. Add multi channel support
3. Implement Access control functionality
4. Introduce plugin compatibility in Jamulus

All of those were already discussed or touched upon somewhere but have implementation/philosophical questions.

Smaller topics were reworking UI components (restructuring how the UI works, redesigning driver/audio device selection, rethinking how the mixer board user information looks like). Usability/Documentation wise, we should streamline the development process and make (installing/)updating servers easier. Infrastructure wise, questions were raised for more reliable directory servers.

While not all of those points are actionable, I think in the next step we should focus on triaging one or two bigger of the feature requests from above. For the smaller topics raised, I think issues should be raised. Please do so!
Thanks all for discussing! I'll open a triaging discussion tomorrow.

**Comment by ann0see** [comment]
Triaging discussion is out: https://github.com/orgs/jamulussoftware/discussions/3382
